kernel/src/utilities/arch_helpers.rs

```diff
@@ -0,0 +1,437 @@
+            *a3 = len as u32;
+        }
+        TRD104SyscallReturn::UserspaceReadableAllowFailure(err, ptr, len) => {
+            *a0 = TRD104SyscallReturnVariant::FailureU32U32 as u32;
+            *a1 = usize::from(err) as u32;
+            *a2 = ptr as u32;
+            *a3 = len as u32;
+        }
+        TRD104SyscallReturn::AllowReadOnlySuccess(ptr, len) => {
+            *a0 = TRD104SyscallReturnVariant::SuccessU32U32 as u32;
+            *a1 = ptr as u32;
+            *a2 = len as u32;
+        }
+        TRD104SyscallReturn::AllowReadOnlyFailure(err, ptr, len) => {
+            *a0 = TRD104SyscallReturnVariant::FailureU32U32 as u32;
+            *a1 = usize::from(err) as u32;
+            *a2 = ptr as u32;
+            *a3 = len as u32;
+        }
+        TRD104SyscallReturn::SubscribeSuccess(ptr, data) => {
+            *a0 = TRD104SyscallReturnVariant::SuccessU32U32 as u32;
+            *a1 = ptr as u32;
+            *a2 = data as u32;
+        }
+        TRD104SyscallReturn::SubscribeFailure(err, ptr, data) => {
+            *a0 = TRD104SyscallReturnVariant::FailureU32U32 as u32;
+            *a1 = usize::from(err) as u32;
+            *a2 = ptr as u32;
+            *a3 = data as u32;
+        }
+        TRD104SyscallReturn::YieldWaitFor(data0, data1, data2) => {
+            *a0 = data0 as u32;
+            *a1 = data1 as u32;
+            *a2 = data2 as u32;
+        }
+    }
+}
+
+/// Enumeration of all system call return type variant identifiers, as a
+/// superset of those defined in [`TRD104`]. This is not yet standardized.
+///
+/// Each variant is associated with the respective variant identifier that would
+/// be passed along with the return value to userspace.
+#[repr(u32)]
+#[derive(Copy, Clone, Debug)]
+pub enum SyscallReturnVariant {
+    Failure = 0,
+    FailureU32 = 1,
+    FailureU32U32 = 2,
+    FailureU64 = 3,
+    FailurePtrUsize = 4,
+    FailurePtrPtr = 5,
+    Success = 128,
+    SuccessU32 = 129,
+    SuccessU32U32 = 130,
+    SuccessU64 = 131,
+    SuccessU32U32U32 = 132,
+    SuccessU32U64 = 133,
+    SuccessUsize = 134,
+    SuccessPtr = 135,
+    SuccessPtrUsize = 136,
+    SuccessPtrPtr = 137,
+}
+
+/// Encode the system call return value into 4 registers, with an encoding that
+/// is a superset of the return variants specified in TRD104. This extension
+/// behaves identical to TRD104 for the [`SyscallReturn`] variants that are
+/// defined in TRD104. This does not implement any standardized encoding.
+///
+/// On 64-bit platforms, both 64-bit and usize values are passed as a single register,
+/// Does not handle usize other than 4 and 8 bytes.
+/// Pointers from allow'd buffers have permissions and length reattached matching
+/// those that were checked at the syscall boundary.
+pub fn encode_syscall_return_usize(
+    syscall_return: &SyscallReturn,
+    a0: &mut CapabilityPtr,
+    a1: &mut CapabilityPtr,
+    a2: &mut CapabilityPtr,
+    a3: &mut CapabilityPtr,
+) {
```

Comment:
This serves to demonstrate how the TRD104 implementation can co-exist with this one. Given it's not used, we may as well add it alongside the PR that introduces the actual CHERI architecture support.

Reply:
So, I have two concerns with this:

The first is that, from an implementation standpoint, there is now clearly code duplication where before there was not. The delta between `encode_syscall_return_usize` and `encode_syscall_return_32bit_trd104` is small, and could be expresed as some kind of parameter (leaving up for discussion if that is a dynamic arg, const parameter, or generic type param that injects some kind of mapping).

The second is that this duplicate interface is likely to lead to duplication in the call sites as well. The `riscv` implementation is going to use one of these functions in one case, and one in the other, adding splits:

```
if [conditions_for_trd104] {
   encode_syscall_return_32bit_trd104(TRD104SyscallReturn::from_syscall_return(syscall), ...)
else
  encode_syscall_return_usize(syscall)
```

As opposed to what it looked like before:

```
encode_syscall_return_usize(syscall)
```

This is duplicated logic, duplicated enums (at both the syscall and variant level), and duplicate calls sites. What specifically are we trying to achieve? Better inject of syscalls mapping choices, rather than have it happen implicitly (as into_compat was doing)? For that, might I suggest:

```
trait VariantMap {
   pub fn map variant (SyscallReturnVariant) -> SyscallReturnVariant;
}

pub fn encode_syscall_return<V : VariantMap>(syscall_return: &SyscallReturn,
    a0: &mut CapabilityPtr,
    a1: &mut CapabilityPtr,
    a2: &mut CapabilityPtr,
    a3: &mut CapabilityPtr,
```

And then we can have

```
struct TRD104; // always 104
struct TRDXXX_COMPAT // new on 64-bit/cheri, 32-bit 104 backwards compat
struct TRDXXX // always with new return codes
```

And the callsite in an arch create would look something like:

```
encode_syscall_return_usize::<TRD104_COMPAT>(syscall)
```

Or even,

```
encode_syscall_return_usize::<PARAM_FROM_BOARD>(syscall)
```

Or, is this about the type system guarantee you have not produced a variant you should not for TRD104? If you want that, sure, you could create the `TRD104SyscallReturnVariant` and map back and forth in the `VariantMap` implementation. But honestly, I would argue this is just logical correctness and easy to verify. Nothing in the type system stops us from returning the wrong TRD104 code, for instance.

Reply:
> I think this is the way to not have to change/replace TRD104. If we don't create these types then we need to update TRD104.

To be clear, nothing here has functionally changed at the user/kernel layer over having the single function vs two. This is an implementation detail. Not sure why that would or would not motivate a change to TRD104, whose remit is the ABI.

> I think I would need to see this to understand..

Hopefully this will all become clear in the 64-bit RISCV PR (coming next). But as a preview: sure, any particular build will either be 32 or 64 bit. But we don't actually have to re-implement an arch crate  just to achieve support for two different XLENs. Just replacing u32 with usize in the RISCV crates in a couple places makes the same code work on 32 _and_ 64 bit (even though it will only be one, for any given run). We can always re-export those from two different crates/functions/etc to name them to be 32/64 bit specific. In the same way that though a number of different cortex crates exist, truly there only needs to be one implementation.

Not having a parameterised encode function makes it harder to write a parameterised arch crate. And XLEN is just another parameter.

Reply:
> Or, is this about the type system guarantee you have not produced a variant you should not for TRD104?

I think this gets closest to the motivation. Now that we have a kernel-internal type, and ABIs that are not necessarily bijective, lossless mappings of the kernel types to an ABI encoding, we want to have an explicit conversion (that we can relatively easily check), and the type-level guarantee that we have actually converted between these types.

While I agree that this is functionally equivalent to what we had before, it eases reasoning about the flow of these values and their conversion between core kernel and arch crate.

> The second is that this duplicate interface is likely to lead to duplication in the call sites as well. The `riscv` implementation is going to use one of these functions in one case, and one in the other, adding splits:

I am largely indifferent about the actual implementation of the conversion method, and whether it's parametrized or duplicated.

I want to emphasize: the version I pushed does not try to be the most elegant, have the least code duplication, or be the most efficient. The version I pushed was supposed to be clearest in demonstrating that the new ABI this (partially) proposes, and the ABI in TRD104 can co-exist, and allow reviewers to reason about this with minimal cognitive overhead.

Anyways, this point is moot now, given that the new `encode_sycall_return_usize` has been removed from this PR. This will be much easier to talk about when we're seeing actual usages of this anyways.